src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+    defer segments_metadata.deinit();
+    const compressed_lines_and_index = mem.bytesAsSlice(f64, compressed_values);
+
+    var general_index: usize = 0;
+    while (general_index < compressed_lines_and_index.len - 1) {
+        const interception: f64 = compressed_lines_and_index[general_index];
+        const slopes_count = @as(usize, @bitCast(compressed_lines_and_index[general_index + 1]));
+        general_index += 2;
+
+        for (0..slopes_count) |_| {
+            const slope = compressed_lines_and_index[general_index];
+            const timestamps_count = @as(usize, @bitCast(compressed_lines_and_index[general_index + 1]));
+            general_index += 2;
+            var timestamp: usize = 0;
+            for (0..timestamps_count) |_| {
+                timestamp += @as(usize, @bitCast(compressed_lines_and_index[general_index]));
+                try segments_metadata.append(.{
+                    .start_time = timestamp,
+                    .interception = interception,
+                    .lower_bound_slope = slope,
+                    .upper_bound_slope = slope,
+                });
+                general_index += 1;
+            }
+        }
+    }
+
+    const last_timestamp: usize = @as(usize, @bitCast(compressed_lines_and_index[general_index]));
+
+    mem.sort(
+        SegmentMetadata,
+        segments_metadata.items,
+        {},
+        compareMetadataByStartTime(),
+    );
+
+    var current_timestamp: usize = 0;
+    for (0..segments_metadata.items.len - 1) |index| {
+        const current_metadata = segments_metadata.items[index];
+        const next_metadata_start_time = segments_metadata.items[index + 1].start_time;
+        while (current_timestamp < next_metadata_start_time) : (current_timestamp += 1) {
+            const approximated_value = current_metadata.upper_bound_slope * @as(f64, @floatFromInt(
+                current_timestamp - current_metadata.start_time,
+            )) + current_metadata.interception;
+            try decompressed_values.append(approximated_value);
+        }
+    }
+
+    const current_metadata = segments_metadata.getLast();
+    while (current_timestamp < last_timestamp) : (current_timestamp += 1) {
+        const approximated_value = current_metadata.upper_bound_slope * @as(f64, @floatFromInt(
+            current_timestamp - current_metadata.start_time,
+        )) + current_metadata.interception;
+        try decompressed_values.append(approximated_value);
+    }
+}
+
+/// Sim-Piece Phase 1: Compute the `segments_metadata` for each segment that can be approximated
+/// by a linear function within the `error_bound` from the `uncompressed_values`.
+fn computeSegmentsMetadata(
+    uncompressed_values: []const f64,
+    segments_metadata: *ArrayList(SegmentMetadata),
+    error_bound: f32,
+) !void {
+    // Adjust the error bound to avoid exceeding it during decompression.
+    const adjusted_error_bound = if (error_bound > 0)
```

Comment:
I could do that, but since this function needs both I will have to pass them as parameters. It think it is better to leave like this since nowhere else the adjusted error bound is computed.